File: lib/android/app/src/main/java/com/reactnativenavigation/presentation/FabPresenter.java
package com.reactnativenavigation.presentation;


import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;

import com.reactnativenavigation.R;
import com.reactnativenavigation.parse.FabOptions;
import com.reactnativenavigation.viewcontrollers.ViewController;
import com.reactnativenavigation.views.Fab;

import androidx.annotation.NonNull;
import androidx.coordinatorlayout.widget.CoordinatorLayout;

import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
import static com.google.android.material.floatingactionbutton.FloatingActionButton.SIZE_MINI;
import static com.google.android.material.floatingactionbutton.FloatingActionButton.SIZE_NORMAL;
import static com.reactnativenavigation.utils.ObjectUtils.perform;

public class FabPresenter {
    private ViewGroup viewGroup;
    private ViewController component;

    private Fab fab;

    public void applyOptions(FabOptions options, @NonNull ViewController component, @NonNull ViewGroup viewGroup) {
        this.viewGroup = viewGroup;
        this.component = component;

        if (options.id.hasValue()) {
            if (fab != null && fab.getFabId().equals(options.id.get())) {
                fab.bringToFront();
                applyFabOptions(fab, options);
                setParams(fab, options);
                fab.setOnClickListener(v -> component.sendOnNavigationButtonPressed(options.id.get()));
            } else {
                createFab(options);
            }
        } else {
            removeFab();
        }
    }

    public void mergeOptions(FabOptions options, @NonNull ViewController component, @NonNull ViewGroup viewGroup) {
        this.viewGroup = viewGroup;
        this.component = component;
        if (options.id.hasValue()) {
            if (fab != null && fab.getFabId().equals(options.id.get())) {
                mergeParams(fab, options);
                fab.bringToFront();
                mergeFabOptions(fab, options);
                fab.setOnClickListener(v -> component.sendOnNavigationButtonPressed(options.id.get()));
            } else {
                createFab(options);
            }
        }
    }

    private void createFab(FabOptions options) {
        fab = new Fab(viewGroup.getContext(), options.id.get());
        setParams(fab, options);
        applyFabOptions(fab, options);
        fab.setOnClickListener(v -> component.sendOnNavigationButtonPressed(options.id.get()));
        viewGroup.addView(fab);
    }

    private void removeFab() {
        if (fab != null) {
            fab.hide();
            viewGroup.removeView(fab);
            fab = null;
        }
    }

    private void setParams(View fab, FabOptions options) {
        CoordinatorLayout.LayoutParams lp = new CoordinatorLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT);
        lp.rightMargin = (int) viewGroup.getContext().getResources().getDimension(R.dimen.margin);
        lp.leftMargin = (int) viewGroup.getContext().getResources().getDimension(R.dimen.margin);
        lp.bottomMargin = (int) viewGroup.getContext().getResources().getDimension(R.dimen.margin);
        fab.setTag(R.id.fab_bottom_margin, lp.bottomMargin);
        lp.gravity = Gravity.BOTTOM;
        if (options.alignHorizontally.hasValue()) {
            if ("right".equals(options.alignHorizontally.get())) {
                lp.gravity |= Gravity.RIGHT;
            }
            if ("left".equals(options.alignHorizontally.get())) {
                lp.gravity |= Gravity.LEFT;
            }
        } else {
            lp.gravity |= Gravity.RIGHT;
        }
        fab.setLayoutParams(lp);
    }

    private void mergeParams(View fab, FabOptions options) {
        CoordinatorLayout.LayoutParams lp = (CoordinatorLayout.LayoutParams) perform(
                fab,
                new CoordinatorLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT),
                View::getLayoutParams
        );
        fab.setTag(R.id.fab_bottom_margin, lp.leftMargin);
        lp.gravity = Gravity.BOTTOM;
        if (options.alignHorizontally.hasValue()) {
            if ("right".equals(options.alignHorizontally.get())) {
                lp.gravity |= Gravity.RIGHT;
            }
            if ("left".equals(options.alignHorizontally.get())) {
                lp.gravity |= Gravity.LEFT;
            }
        } else {
            lp.gravity |= Gravity.RIGHT;
        }
        fab.setLayoutParams(lp);
    }

    private void applyFabOptions(Fab fab, FabOptions options) {
        if (options.visible.isTrueOrUndefined()) {
            fab.show();
        }
        if (options.visible.isFalse()) {
            fab.hide();
        }
        if (options.backgroundColor.hasValue()) {
            fab.setBackgroundColor(options.backgroundColor.get());
        }
        if (options.clickColor.hasValue()) {
            fab.setRippleColor(options.clickColor.get());
        }
        if (options.icon.hasValue()) {
            fab.applyIcon(options.icon.get(), options.iconColor);
        }
        if (options.size.hasValue()) {
            fab.setSize("mini".equals(options.size.get()) ? SIZE_MINI : SIZE_NORMAL);
        }
        if (options.hideOnScroll.isTrue() && options.visible.isTrueOrUndefined()) {
            fab.enableCollapse(component.getScrollEventListener());
        }
        if (options.hideOnScroll.isFalseOrUndefined()) {
            fab.disableCollapse();
        }
        if (options.customSize.hasValue() && options.customSize.get() >= 50) {
            fab.setCustomSize(options.customSize.get());
        }
    }

    private void mergeFabOptions(Fab fab, FabOptions options) {
        if (options.visible.isTrue()) {
            fab.show();
        }
        if (options.visible.isFalse()) {
            fab.hide();
        }
        if (options.backgroundColor.hasValue()) {
            fab.setBackgroundColor(options.backgroundColor.get());
        }
        if (options.clickColor.hasValue()) {
            fab.setRippleColor(options.clickColor.get());
        }
        if (options.icon.hasValue()) {
            fab.applyIcon(options.icon.get(), options.iconColor);
        }
        if (options.size.hasValue()) {
            fab.setSize("mini".equals(options.size.get()) ? SIZE_MINI : SIZE_NORMAL);
        }
        if (options.hideOnScroll.isTrue() && options.visible.isTrueOrUndefined()) {
            fab.enableCollapse(component.getScrollEventListener());
        }
        if (options.hideOnScroll.isFalse()) {
            fab.disableCollapse();
        }
        if (options.customSize.hasValue() && options.customSize.get() >= 50) {
            fab.setCustomSize(options.customSize.get());
        }
    }
}

File: lib/android/app/src/main/java/com/reactnativenavigation/presentation/FabPresenter.kt
package com.reactnativenavigation.presentation

import android.content.res.ColorStateList
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import androidx.coordinatorlayout.widget.CoordinatorLayout
import com.google.android.material.floatingactionbutton.FloatingActionButton.SIZE_MINI
import com.google.android.material.floatingactionbutton.FloatingActionButton.SIZE_NORMAL
import com.reactnativenavigation.R
import com.reactnativenavigation.interfaces.FloatingButton
import com.reactnativenavigation.parse.FabOptions
import com.reactnativenavigation.parse.Options
import com.reactnativenavigation.utils.ObjectUtils
import com.reactnativenavigation.viewcontrollers.ViewController
import com.reactnativenavigation.views.ExtendedFab
import com.reactnativenavigation.views.Fab


class FabPresenter {
    private var viewGroup: ViewGroup? = null
    private var component: ViewController<*>? = null
    private var fab:FloatingButton? = null

    private fun createFab(options: Options) {
        if (options.extendedFabOptions.hasValue()) {
            fab = ExtendedFab(viewGroup!!.context, options.fabOptionsType.id.get())
        } else {
            fab = Fab(viewGroup!!.context, options.fabOptionsType.id.get())
        }
        with(options.fabOptionsType) {
            setParams(fab!! as View, this)
            applyFabOptions(fab!!, this)
            fab!!.setOnClickListener { _: View? -> component!!.sendOnNavigationButtonPressed(this.id.get()) }
        }

        viewGroup!!.addView(fab as View)
    }

    fun applyOptions(options: Options, component: ViewController<*>, viewGroup: ViewGroup) {
        this.viewGroup = viewGroup
        this.component = component
        with(options.fabOptionsType) {
            if (this.id.hasValue()) {
                if (fab != null && fab!!.fabId == this.id.get()) {
                    fab!!.bringToFront()
                    applyFabOptions(fab!!, this)
                    setParams(fab!! as View, this)
                    fab!!.setOnClickListener { _: View? -> component.sendOnNavigationButtonPressed(this.id.get()) }
                } else {
                    createFab(options)
                }
            } else {
                removeFab()
            }
        }
    }

    fun mergeOptions(options: Options, component: ViewController<*>, viewGroup: ViewGroup) {
        this.viewGroup = viewGroup
        this.component = component
        with(options.fabOptionsType) {
            if (this.id.hasValue()) {
                if (fab != null && fab!!.fabId == this.id.get()) {
                    mergeParams(fab!! as View, this)
                    fab!!.bringToFront()
                    mergeFabOptions(fab!!, this)
                    fab!!.setOnClickListener { _: View? -> component.sendOnNavigationButtonPressed(this.id.get()) }
                } else {
                    createFab(options)
                }
            }
        }
    }

    private fun removeFab() {
        if (fab != null) {
            fab!!.hide()
            viewGroup!!.removeView(fab as View)
            fab = null
        }
    }

    private fun setParams(fab: View, options: FabOptions) {
        val lp = CoordinatorLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        lp.rightMargin = viewGroup!!.context.resources.getDimension(R.dimen.margin).toInt()
        lp.leftMargin = viewGroup!!.context.resources.getDimension(R.dimen.margin).toInt()
        lp.bottomMargin = viewGroup!!.context.resources.getDimension(R.dimen.margin).toInt()
        fab.setTag(R.id.fab_bottom_margin, lp.bottomMargin)
        lp.gravity = Gravity.BOTTOM
        if (options.alignHorizontally.hasValue()) {
            if ("right" == options.alignHorizontally.get()) {
                lp.gravity = lp.gravity or Gravity.RIGHT
            }
            if ("left" == options.alignHorizontally.get()) {
                lp.gravity = lp.gravity or Gravity.LEFT
            }
        } else {
            lp.gravity = lp.gravity or Gravity.RIGHT
        }
        fab.layoutParams = lp
    }

    private fun mergeParams(fab: View, options: FabOptions) {
        val lp = ObjectUtils.perform(
                fab,
                CoordinatorLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT), { obj: View -> obj.layoutParams }) as CoordinatorLayout.LayoutParams
        fab.setTag(R.id.fab_bottom_margin, lp.leftMargin)
        lp.gravity = Gravity.BOTTOM
        if (options.alignHorizontally.hasValue()) {
            if ("right" == options.alignHorizontally.get()) {
                lp.gravity = lp.gravity or Gravity.RIGHT
            }
            if ("left" == options.alignHorizontally.get()) {
                lp.gravity = lp.gravity or Gravity.LEFT
            }
        } else {
            lp.gravity = lp.gravity or Gravity.RIGHT
        }
        fab.layoutParams = lp
    }

    private fun applyFabOptions(fab: FloatingButton, options: FabOptions) {
        if (options.visible.isTrueOrUndefined) {
            fab.show()
        }
        if (options.visible.isFalse) {
            fab.hide()
        }
        if (options.backgroundColor.hasValue()) {
            fab.setBackgroundColor(options.backgroundColor.get())
        }
        if (options.clickColor.hasValue()) {
            fab.setRippleColor(ColorStateList.valueOf(options.clickColor.get()))
        }
        if (options.icon.hasValue()) {
            fab.applyIcon(options.icon.get(), options.iconColor)
        }
        if (options.hideOnScroll.isTrue && options.visible.isTrueOrUndefined) {
            fab.enableCollapse(component!!.scrollEventListener)
        }
        if (options.hideOnScroll.isFalseOrUndefined) {
            fab.disableCollapse()
        }
        if (options.size.hasValue()) {
            fab.setSize(if ("mini" == options.size.get()) SIZE_MINI else SIZE_NORMAL)
        }
        if (options.customSize.hasValue() && options.customSize.get() >= 50) {
            fab.setCustomSize(options.customSize.get())
        }
        if (options.text.hasValue()) {
            fab.setText(options.text.get())
        }
        if (options.setExtended.hasValue()) {
            fab.setExtended(options.setExtended.get())
        }
    }

    private fun mergeFabOptions(fab: FloatingButton, options: FabOptions) {
        if (options.visible.isTrue) {
            fab.show()
        }
        if (options.visible.isFalse) {
            fab.hide()
        }
        if (options.backgroundColor.hasValue()) {
            fab.setBackgroundColor(options.backgroundColor.get())
        }
        if (options.clickColor.hasValue()) {
            fab.setRippleColor(ColorStateList.valueOf(options.clickColor.get()))
        }
        if (options.icon.hasValue()) {
            fab.applyIcon(options.icon.get(), options.iconColor)
        }
        if (options.hideOnScroll.isTrue && options.visible.isTrueOrUndefined) {
            fab.enableCollapse(component!!.scrollEventListener)
        }
        if (options.hideOnScroll.isFalse) {
            fab.disableCollapse()
        }
        if (options.size.hasValue()) {
            fab.setSize(if ("mini" == options.size.get()) SIZE_MINI else SIZE_NORMAL)
        }
        if (options.customSize.hasValue() && options.customSize.get() >= 50) {
            fab.setCustomSize(options.customSize.get())
        }
        if (options.text.hasValue()) {
            fab.setText(options.text.get())
        }
        if (options.setExtended.hasValue()) {
            fab.setExtended(options.setExtended.get())
        }
    }
}